java/ql/lib/semmle/code/java/security/PathSanitizer.qll

```diff
@@ -123,7 +122,7 @@ private predicate blockListGuard(Guard g, Expr e, boolean branch) {
   )
 }
 
-private class BlockListSanitizer extends PathTraversalSanitizer {
+private class BlockListSanitizer extends PathInjectionSanitizer {
   BlockListSanitizer() { this = DataFlow::BarrierGuard<blockListGuard/3>::getABarrierNode() }
```

Comment:
For the `blockListGuard` we require that the user has mitigated for URL decoding issues. Is this really a universal requirement to use a block list?

As far as I know, the only impact URI decoding has on tainted path issues is that decoding should not occur after path checks, because the path may only be malicious after decoding.

However, requiring the user _decode_ the path before checking can be outright harmful if using a library or framework that already URI decodes remote flow sources, or for remote flow sources that don't require URI decoding.

Reply:
In general, block lists are less than ideal sanitizers, and allow lists should be preferred. One of the reasons for that is that it's easy to forget about a bad case, or to not consider different encodings when checking the externally-provided data.

So this condition is there to force block lists approaches to be a little more robust if we want them to be considered valid sanitizers. I agree that the requirement of URL decoding prior to the check might be too much, but theoretically your examples:

> if using a library or framework that already URI decodes remote flow sources, or for remote flow sources that don't require URI decoding.

could be fixed by marking those specific remote flow sources as `UrlDecoderSanitizer`s themselves.

The alternative would be that a simple `payload.contains("..")` would be considered a sanitizer, and then users would complain as well, since we all know that that doesn't work at all. Again, at some point we need to trade FNs for FPs, or vice versa.

Reply:
> The alternative would be that a simple payload.contains("..") would be considered a sanitizer

I agree that we don't want that, but I don't think requiring a URI decoding step beforehand makes `payload.contains("..")` a sufficient sanitizer for absolute path traversal, for the same reasons as above.

> could be fixed by marking those specific remote flow sources as UrlDecoderSanitizers themselves.

I don't think requiring URI decoding _before_ the check prevents any security issues. The URI decoding issue is only a problem if you URI decode _between_ the check and the access, e.g.
```
x = decode(source())
if (check(x)) {
  new File(x); // fine
}

x = source()
if (check(x)) {
  new File(x); // also fine, File won't URI decode
}

x = source()
if (check(x)) {
  new File(decode(x)); // bad!
}
```

For example, in the Spring vulnerability the issue wasn't that they forgot to URI decode before their tainted path check. Instead, the problem was that they ran second decode step _after_ the check, but before the value was used.

Reply:
Clearly, requiring taking double-encoding into account only makes sense if the sink itself implicitly URL-decodes the value, which happens in some dispatch/forward/redirect vulnerabilities. I think this could actually be another interesting use case for flow states (see the comment I added in the now FN test in https://github.com/github/codeql/pull/10177/files#diff-6bc44c5765a6e460db430d2c13a69b22afe818b46bc96d630508b5e1d1cd6eb1R103-R106).

So, with that in mind, I followed your advice and removed the URL-decode requirement for `blockListGuard`. Instead, I added the same canonicalization/path traversal check requirement that `allowListGuard` has, assuming that the block list will check for dangerous prefixes (like `/data` in Android). The only case that would not need normalization would be `contains(File.seperator)` (or `contains("/")` or `contains("\\")`), but to special-case those bad words we would need to create yet another guard. We can discuss if you think that'd be worth it.